Python/ceval.c

```diff
@@ -2119,9 +2128,11 @@ _PyEval_UnpackIterableStackRef(PyThreadState *tstate, _PyStackRef v_stackref,
             return 1;
         }
         Py_DECREF(w);
-        _PyErr_Format(tstate, PyExc_ValueError,
-                      "too many values to unpack (expected %d)",
-                      argcnt);
+        unpack_args = Py_BuildValue("OOi", Py_None, v, argcnt);
+        if (unpack_args) {
```

Comment:
Here, what you could do is actually compute the length so that you don't need to store the iterable. If computing the iterable fails, then you will just propagate the exception normally *or* you could fallback to the previous message. To check whether `__len__` exists, you may first use `_PyObject_HasLen` and then call it. Then, you call `PyObject_Size` and check if the result is ok or not. If not, that means there is an exception in `__len__` (and not that the object was not lengthable) which must be propagated as is. For an object that does not have `__len__`, you would however fallback to the original message.

This will require you to change the tests though.

Reply:
This sounds like what I was trying to do in my original attempt, ref https://github.com/python/cpython/pull/122244/files/cd0d534a7040ecb5eb0c0a39a2e03c7550f07cba

I am fine with this approach as well, but was suggested to move the error message changes to `traceback.py` earlier. @encukou what do you think of this suggestion? I do agree this is much simpler than inventing a new top level error class.